src/ray/common/client_connection.h

```diff
@@ -94,7 +94,7 @@ class ServerConnection : public std::enable_shared_from_this<ServerConnection<T>
   const int async_write_max_messages_;
 
   /// List of pending messages to write.
-  std::list<std::unique_ptr<AsyncWriteBuffer>> async_write_queue_;
+  std::deque<std::unique_ptr<AsyncWriteBuffer>> async_write_queue_;
```

Comment:
Just a note, there's a much more lightweight substitute to `deque` in most cases like this: just use `vector` along with a `size_t` offset indicating the beginning of the queue. When the offset grows bigger than half the size of the whole queue then drop everything before it. You can abstract this away into a class if you'd like. It should be faster both for compiling as well as running.

Reply:
It depends a lot on the implementation but I believe it's generally pretty complicated. IIRC many if not all of them manage lots of separate chunks of memory (some details are [here](https://stackoverflow.com/questions/6292332)... which I don't know in full detail). There is [`stable_vector` in Boost](https://www.boost.org/doc/libs/1_66_0/doc/html/container/non_standard_containers.html#container.non_standard_containers.stable_vector) which is probably closer to the implementation you have in mind. I've always ended up finding `deque` to not be particularly good at anything and slowing things down too much, though I haven't dealt with every implementation.

Reply:
I'm a huge fan of deque -- it's like a chunked vector that supports O(1) operations at both ends, with less memory overhead than std::list. It sometimes can be kind of as bad as list depending on the implementation though.

Reply:
I agree it's better than `list`, that's why I avoid `list` first and `deque` second when :) but have you tried the `vector` implementation I suggested above before to see how it compares? It also has O(1) operations at both ends and the overhead is just like that of a `vector`, and it should compile faster as well.

Reply:
@mehrdadn , it's better to optimize for readable code than worry about slight inefficiencies of `std::deque`. In my experience, deque is very high performing and you end up allocating chunks of the same size which can end up performing better in practice. Using two variables and custom logic is just over-optimization and likely to introduce bugs. It's just not worth the implementation cost.

Reply:
Well that's why I said you can abstract it away into a class if you'd like (just like you already did with `ordered_set`). In terms of its performance though I'm unclear on whether, when you said it can end up performing better in your experience, you meant compared to the performance you observed with the `vector` implementation above, or with something else. But in any case, you obviously don't have to change it if you don't want to.